### ISSUE TITLE
[RFE] Add support for Azure storage type StandardSSD_LRS

### ISSUE THREAD

**Issue by gaktive** [rancher/dashboard#7177]
Internal reference: SURE-3116
Reported in 2.5.8

There's a request to add the ability to select storage types when creating or editing node templates for Azure.

Requested choices:

StandardSSD LRS 
https://azure.microsoft.com/fr-fr/blog/preview-standard-ssd-disks-for-azure-virtual-machine-workloads/

UltraSSD LRS (this won't be possible yet for reasons listed in the comments)
https://azure.microsoft.com/nl-nl/blog/announcing-the-general-availability-of-azure-ultra-disk-storage/ 

Actual behavior:
The disk choices aren't available

Expected behavior:
The disk choices should be available

@HarrisonWAffel had mentioned this in the midst of research, though no ticket found in rancher/rancher repo:

> Can you confirm if the user is asking us to enhance the list of currently available storage types when creating an Azure VM (typically called an OS disk), or if they are asking us to provide support for creating and managing Data-disks within node templates? 

> Data-disks are additional storage options that do not contain the OS. I only ask because the UltraSSD LRS option cannot be used as a OS disk for an Azure VM and can only be used as a data-disk. This is detailed in the Azure documentation https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-ultra-ssd?tabs=azure-portal#ga-scope-and-limitations. Currently the storage type option within the node template for Azure refers to the OS disk used by the VM. The StandardSSD LRS option can be added to the current node template, but currently there is no way to manage data-disks for Azure VM's from within Rancher. If they are only asking us to expand the option for OS disks then the scope of work is quite small.

> Assuming that the user is requesting an enhancement to the current list of Azure VM OS storage disks, I believe that changes would be localized to the Rancher dashboard. I've done my best to determine the required changes and story point estimates, however the UI team should have final say. 

### Dashboard Changes (0.5 pt, need confirmation with UI team) 
- The storage types defined in `dashboard/shell/machine-config/azure.vue` would need to be updated to include a new entry
-- It should be named `Standard SSD LRS` and have a value of `StandardSSD_LRS` 
- The storage type `UltraSSD LRS` cannot be added to the dashboard for Azure VM OS disks, as Ultra-SSD's can only be used a data-disks. At the moment, Rancher does not provide a means of creating or managing data-disks for Azure VM's. The documentation listing the limitations of Ultra-disks can be found here: https://learn.microsoft.com/en-us/azure/virtual-machines/disks-types#ultra-disk-limitations
 
(Still need to confirm if the request is for us to "implement support for Azure Data-disk management within Rancher" as that would require changes.)

**Comment by catherineluse** [comment]
It looks like most of the Azure related changes are related to node templates, which only apply to RKE1 clusters. Are these options just for RKE1 node templates, or do they also need to be added to the node pool forms for RKE2 provisioning?

![Screen Shot 2022-10-20 at 11.57.59 AM.png](https://zube.io/files/rancher/2c1117f038a6645a89a3814c7aee89ca-screen-shot-2022-10-20-at-11-57-59-am.png)

**Comment by catherineluse** [comment]
While we're waiting for an answer about Ultradisk, I've gone ahead and added StandardSSD as an option:

- RKE1: https://github.com/rancher/ui/pull/4905
- RKE2/K3s: https://github.com/rancher/dashboard/pull/7254

Since Ultradisk would be more complicated to add, should it be tracked in a separate issue?

Side note: It would also be a little more complicated to add Ultradisk on the UI side due to the fact that Ultradisk is not available in all regions. Ideally, if the user has selected a region where the option is not available, I'd like that option to be disabled with a tooltip saying "Not available in the selected region." (That's not needed for StandardSSD because it's available in all regions.)

**Comment by gaktive** [comment]
@catherineluse I checked with @HarrisonWAffel and it's a 'will not do':
> UltraDisks are considered 'data-disks' and right now there isn't a way to manage those in Rancher. The reporters latest comments indicate to me that they aren't asking us to support them so I think that disk type can be disregarded.

This may change down the road, of course, but for now, let's do the minimum and when we get a request, we can file a ticket later to add UltraDisks.

**Comment by gaktive** [comment]
Waiting on https://github.com/rancher/machine/pull/198 to land to make this workable. As noted, Ultradisk won't work so we'll explore a warning in case someone tries to pick it (or we just hide it with an explanation). As to whether future storage types run into this issue, then we may have to look at some other futureproofing.

**Comment by gaktive** [comment]
Backend complete, so they're awaiting UI before handing to QA: https://github.com/rancher/rancher/issues/39672 though they will await for all related Azure AD tickets to be completed before testing begins.

**Comment by catherineluse** [comment]
# QA Template
## What was fixed, or what changes have occurred
Added the StandardSSD_LRS storage option for RKE1 and RKE2/K3s Azure clusters 

 
## Areas or cases that should be tested

- RKE1 Azure clusters - see screenshot https://github.com/rancher/ui/pull/4905
- RKE2/K3s Azure clusters - see screenshot https://github.com/rancher/dashboard/pull/7254

**Comment by snasovich** [comment]
@gaktive @catherineluse , just to clarify something. https://github.com/rancher/dashboard/issues/7177#issuecomment-1317761114 mentions Rancher machine changes are needed for this, but from what I see StorageType was available prior to the change and this issue is about adding an extra option for this field. Correct?

**Comment by thaneunsoo** [comment]
I was able to verify that the node template was created successfully but was not able to test cluster provisioning as I am blocked by [7636](https://github.com/rancher/dashboard/issues/7636) and will validate the cluster provisioning once the issue is fixed.
![image.png](https://zube.io/files/rancher/e8d81c1b06b6388b43035aa1a6dfb9f7-image.png)

Moving to QA blocked until blocker is resolved

**Comment by thaneunsoo** [comment]
### Test Environment: ###
Rancher version: v2.7-head a2ef50b
Rancher cluster type: HA
Docker version: 20.10

Downstream cluster type: rke1 & rke2 azure node driver

----------

### Testing: ###

I seem to have found an issue with the `StandardSSD_LRS` selection for both rke1 and rke2. The rest of the dropdown list omits the `SSD` part in the name and when I checked the [azure site](https://learn.microsoft.com/en-us/rest/api/storagerp/storage-accounts/create?tabs=HTTP#skuname) where they list the SkuName, I saw that there did not seem to be an option for Standard_SSD_LRS
![image.png](https://zube.io/files/rancher/e5ec64025fb631947bbcb10ceff87460-image.png)
When trying to create a cluster using the node template I get the following error for rke1 and rke2
RKE1
```
Error creating machine: Error in driver during machine creation: storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidAccountType" Message="AccountType StandardSSD_LRS is invalid. For more information, see - https://aka.ms/storageaccountskus":Timeout waiting for ssh key
```

RKE2
```
Error creating machine: Error in driver during machine creation:
storage.AccountsClient#Create: Failure sending request: StatusCode=0 --
Original Error: Code="InvalidAccountType" Message="AccountType
StandardSSD_LRS is invalid. For more information, see -
https://aka.ms/storageaccountskus"
```


Here are the options in the dropdown list for both rke1 and rke2
RKE2
![image.png](https://zube.io/files/rancher/3c846ae58370d8082cdaf0b1aafa5a83-image.png)
RKE1
![image.png](https://zube.io/files/rancher/e8d81c1b06b6388b43035aa1a6dfb9f7-image.png)

@catherineluse could you take a look? thanks!

**Comment by HarrisonWAffel** [comment]
@thaneunsoo `StandardSSD_LRS` requires you use a `Managed Disk` type, in your screenshot it looks like you used unmanaged disks. Though, we should probably have a UI hint as to this requirement, since the error returned from Azure is not very informative.

**Comment by catherineluse** [comment]
I created separate PRs that show an error message/UI hint if StandardSSD_LRL is selected without managed disks enabled:

- RKE2/K3s: https://github.com/rancher/dashboard/issues/7743
- RKE1: https://github.com/rancher/dashboard/issues/7744

I created separate issues linked to each of them so that this one can be closed out.

**Comment by thaneunsoo** [comment]
Closing this issue based on below comments.
